File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_cpp/ZipFileProcessor.cpp
#include <zip.h>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <memory>
#include <vector>
#include <string>
#include <stdexcept>
#include <cstring>

namespace fs = std::filesystem;

enum class Level {
    SEVERE,
    WARNING,
    INFO
};

class Logger {
public:
    Logger(const std::string& name) : name_(name) {}

    void log(Level level, const std::string& msg) {
        const char* levelStr = "";
        switch (level) {
            case Level::SEVERE: levelStr = "SEVERE"; break;
            case Level::WARNING: levelStr = "WARNING"; break;
            case Level::INFO: levelStr = "INFO"; break;
        }
        std::cerr << "[" << name_ << "] " << levelStr << ": " << msg << std::endl;
    }

    void log(Level level, const std::string& msg, const std::exception& e) {
        log(level, msg + ": " + e.what());
    }

    void severe(const std::string& msg) { log(Level::SEVERE, msg); }
    void warning(const std::string& msg) { log(Level::WARNING, msg); }
    void info(const std::string& msg) { log(Level::INFO, msg); }

    static Logger getLogger(const std::string& name) {
        return Logger(name);
    }

private:
    std::string name_;
};

struct ZipEntry {
    std::string name;
    bool isDirectory;
};

class ZipFile {
public:
    explicit ZipFile(zip_t* handle) : handle_(handle) {}

    ~ZipFile() {
        if (handle_) {
            zip_close(handle_);
        }
    }

    ZipFile(const ZipFile&) = delete;
    ZipFile& operator=(const ZipFile&) = delete;

    ZipFile(ZipFile&& other) noexcept : handle_(other.handle_) {
        other.handle_ = nullptr;
    }

    ZipFile& operator=(ZipFile&& other) noexcept {
        if (this != &other) {
            if (handle_) {
                zip_close(handle_);
            }
            handle_ = other.handle_;
            other.handle_ = nullptr;
        }
        return *this;
    }

    bool isValid() const { return handle_ != nullptr; }

    std::vector<ZipEntry> entries() {
        std::vector<ZipEntry> result;
        zip_int64_t num_entries = zip_get_num_entries(handle_, 0);
        for (zip_int64_t i = 0; i < num_entries; i++) {
            struct zip_stat sb;
            if (zip_stat_index(handle_, i, 0, &sb) != 0) {
                continue;
            }
            std::string name = sb.name;
            bool isDirectory = (sb.valid & ZIP_STAT_NAME) && name.back() == '/';
            result.push_back({name, isDirectory});
        }
        return result;
    }

    std::optional<ZipEntry> getEntry(const std::string& name) {
        zip_int64_t index = zip_name_locate(handle_, name.c_str(), 0);
        if (index < 0) {
            return std::nullopt;
        }
        struct zip_stat sb;
        if (zip_stat_index(handle_, index, 0, &sb) != 0) {
            return std::nullopt;
        }
        std::string entryName = sb.name;
        bool isDirectory = (sb.valid & ZIP_STAT_NAME) && entryName.back() == '/';
        return ZipEntry{entryName, isDirectory};
    }

    bool extractEntry(const ZipEntry& entry, const fs::path& filePath) {
        zip_int64_t index = zip_name_locate(handle_, entry.name.c_str(), 0);
        if (index < 0) {
            return false;
        }

        zip_file_t* file = zip_fopen_index(handle_, index, 0);
        if (!file) {
            return false;
        }

        try {
            std::ofstream out(filePath, std::ios::binary | std::ios::trunc);
            if (!out) {
                zip_fclose(file);
                return false;
            }

            char buffer[1024];
            zip_int64_t len;
            while ((len = zip_fread(file, buffer, sizeof(buffer))) {
                if (len < 0) {
                    zip_fclose(file);
                    return false;
                }
                out.write(buffer, len);
                if (!out) {
                    zip_fclose(file);
                    return false;
                }
            }
            zip_fclose(file);
            return true;
        } catch (...) {
            zip_fclose(file);
            return false;
        }
    }

private:
    zip_t* handle_;
};

class ZipFileProcessor {
private:
    fs::path zipFilePath;
    static Logger logger;

public:
    ZipFileProcessor(const std::string& zipFileName) : zipFilePath(zipFileName) {}

    std::unique_ptr<ZipFile> readZipFile() {
        if (!fs::exists(zipFilePath)) {
            logger.severe("Zip file does not exist: " + zipFilePath.string());
            return nullptr;
        }

        int err;
        zip_t* handle = zip_open(zipFilePath.c_str(), ZIP_RDONLY, &err);
        if (!handle) {
            char error_str[100];
            zip_error_to_str(error_str, sizeof(error_str), err, errno);
            logger.severe("Failed to read the zip file: " + zipFilePath.string() + " - " + error_str);
            return nullptr;
        }
        return std::make_unique<ZipFile>(handle);
    }

    bool extract_all(const std::string& outputDir) {
        fs::path outputDirPath(outputDir);
        auto zipFile = readZipFile();
        if (!zipFile || !zipFile->isValid()) {
            return false;
        }

        if (!fs::exists(outputDirPath)) {
            try {
                fs::create_directories(outputDirPath);
                logger.info("Created output directory: " + outputDirPath.string());
            } catch (const std::exception& e) {
                logger.log(Level::SEVERE, "Failed to create output directory: " + outputDirPath.string(), e);
                return false;
            }
        }

        auto entries = zipFile->entries();
        for (const auto& entry : entries) {
            fs::path filePath = outputDirPath / entry.name;
            try {
                if (entry.isDirectory) {
                    if (!fs::exists(filePath)) {
                        fs::create_directories(filePath);
                        logger.info("Created directory: " + filePath.string());
                    }
                } else {
                    fs::path parentPath = filePath.parent_path();
                    if (!fs::exists(parentPath)) {
                        fs::create_directories(parentPath);
                    }
                    if (zipFile->extractEntry(entry, filePath)) {
                        logger.info("Extracted file: " + filePath.string());
                    } else {
                        logger.severe("Failed to extract file: " + filePath.string());
                    }
                }
            } catch (const std::exception& e) {
                logger.log(Level::SEVERE, "Failed to process entry: " + entry.name, e);
            }
        }
        return true;
    }

    bool extract_file(const std::string& fileName, const std::string& outputDir) {
        fs::path outputDirPath(outputDir);
        fs::path filePath = outputDirPath / fileName;
        auto zipFile = readZipFile();
        if (!zipFile || !zipFile->isValid()) {
            return false;
        }

        auto entry = zipFile->getEntry(fileName);
        if (!entry) {
            logger.warning("File not found in zip: " + fileName);
            return false;
        }

        fs::path parentPath = filePath.parent_path();
        if (!fs::exists(parentPath)) {
            try {
                fs::create_directories(parentPath);
                logger.info("Created parent directory: " + parentPath.string());
            } catch (const std::exception& e) {
                logger.log(Level::SEVERE, "Failed to create parent directory: " + parentPath.string(), e);
                return false;
            }
        }

        if (zipFile->extractEntry(*entry, filePath)) {
            logger.info("Extracted file: " + filePath.string());
            return true;
        } else {
            logger.severe("Failed to extract file: " + fileName);
            return false;
        }
    }
};

Logger ZipFileProcessor::logger = Logger::getLogger("ZipFileProcessor");

File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/ZipFileProcessor.py
import logging
import zipfile
import io
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)


class ZipFileProcessor:
    def __init__(self, zip_file_name: str):
        self.zip_file_path = Path(zip_file_name)

    def read_zip_file(self) -> Optional[zipfile.ZipFile]:
        if not self.zip_file_path.exists():
            logger.error(
                f"Zip file does not exist: {self.zip_file_path.resolve()}"
            )
            return None
        try:
            return zipfile.ZipFile(str(self.zip_file_path), "r")
        except (OSError, zipfile.BadZipFile) as e:
            logger.exception(
                f"Failed to read the zip file: {self.zip_file_path.resolve()}"
            )
            return None

    def extract_all(self, output_dir: str) -> bool:
        output_dir_path = Path(output_dir)
        zip_file = self.read_zip_file()
        if zip_file is None:
            return False
        try:
            if not output_dir_path.exists():
                try:
                    output_dir_path.mkdir(parents=True, exist_ok=False)
                    logger.info(
                        f"Created output directory: {output_dir_path.resolve()}"
                    )
                except OSError as e:
                    logger.exception(
                        f"Failed to create output directory: {output_dir_path.resolve()}"
                    )
                    return False

            for entry in zip_file.infolist():
                file_path = output_dir_path / entry.filename
                try:
                    if entry.filename.endswith("/"):
                        if not file_path.exists():
                            file_path.mkdir(parents=True, exist_ok=True)
                            logger.info(
                                f"Created directory: {file_path.resolve()}"
                            )
                    else:
                        # Write the file
                        with zip_file.open(entry, "r") as in_file, open(
                            file_path, "wb"
                        ) as out_file:
                            buffer = bytearray(1024)
                            while True:
                                n = in_file.readinto(buffer)
                                if n == 0:
                                    break
                                out_file.write(buffer[:n])
                        logger.info(
                            f"Extracted file: {file_path.resolve()}"
                        )
                except (OSError, zipfile.BadZipFile) as e:
                    logger.exception(
                        f"Failed to process entry: {entry.filename}"
                    )
            return True
        finally:
            zip_file.close()

    def extract_file(self, file_name: str, output_dir: str) -> bool:
        output_dir_path = Path(output_dir)
        file_path = output_dir_path / file_name
        zip_file = self.read_zip_file()
        if zip_file is None:
            return False
        try:
            if file_name not in zip_file.namelist():
                logger.warning(f"File not found in zip: {file_name}")
                return False

            entry = zip_file.getinfo(file_name)

            parent = file_path.parent
            if not parent.exists():
                try:
                    parent.mkdir(parents=True, exist_ok=False)
                    logger.info(
                        f"Created parent directory: {parent.resolve()}"
                    )
                except OSError as e:
                    logger.exception(
                        f"Failed to create parent directory: {parent.resolve()}"
                    )
                    return False

            with zip_file.open(entry, "r") as in_file, open(
                file_path, "wb"
            ) as out_file:
                buffer = bytearray(1024)
                while True:
                    n = in_file.readinto(buffer)
                    if n == 0:
                        break
                    out_file.write(buffer[:n])
            logger.info(f"Extracted file: {file_path.resolve()}")
            return True
        except (OSError, zipfile.BadZipFile) as e:
            logger.exception(f"Failed to extract file: {file_name}")
            return False
        finally:
            zip_file.close()